import re
import asyncio
import grpc
import logging
from aiogram import Router, F, Bot
from aiogram.types import InlineKeyboardButton, Message, CallbackQuery, ReactionTypeEmoji, FSInputFile
from aiogram.fsm.context import FSMContext
from aiogram.utils.keyboard import InlineKeyboardBuilder
from internal.pb.python import downloader_pb2, downloader_pb2_grpc
from internal.bot.config.config import GRPC_DOWNLOADER_ADDRESS
from internal.bot.client.client import send_file_to_bot
from internal.utils.utils import check_path, delete_file

router = Router(name=__name__)

async def get_available_qualities(link: str) -> list:
    """
    –ó–∞–ø—É—Å–∫–∞–µ—Ç yt-dlp –¥–ª—è –ø–æ–ª—É—á–µ–Ω–∏—è –¥–æ—Å—Ç—É–ø–Ω—ã—Ö –∫–∞—á–µ—Å—Ç–≤ –≤–∏–¥–µ–æ.
    –ò—Å–∫–ª—é—á–∞–µ—Ç –∫–∞—á–µ—Å—Ç–≤–∞ –≤—ã—à–µ 1080p.
    –ï—Å–ª–∏ –Ω–µ —É–¥–∞—ë—Ç—Å—è –æ–ø—Ä–µ–¥–µ–ª–∏—Ç—å, –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç –∑–Ω–∞—á–µ–Ω–∏–µ –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é.
    """
    try:
        process = await asyncio.create_subprocess_exec(
            "yt-dlp", "-F", link,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE
        )
        stdout, _ = await process.communicate()

        qualities = set()
        for line in stdout.decode().splitlines():
            # –ò—â–µ–º —Å—Ç—Ä–æ–∫–∏ —Å —Ä–∞–∑—Ä–µ—à–µ–Ω–∏–µ–º –≤–∏–¥–∞ 360p, 480p, 720p –∏ —Ç.–¥.
            match = re.search(r"(\d{3,4}p)", line)
            if match:
                qualities.add(match.group(1))

        if qualities:
            # –ü—Ä–µ–æ–±—Ä–∞–∑—É–µ–º –≤ —Å–ø–∏—Å–æ–∫ –∏ —Ñ–∏–ª—å—Ç—Ä—É–µ–º: —Ç–æ–ª—å–∫–æ <= 1080p
            valid_qualities = [q for q in qualities if int(q.replace("p", "")) <= 1080]
            return sorted(valid_qualities, key=lambda x: int(x.replace("p", "")))
    except Exception as e:
        pass

    return ["720p"]  # –∑–Ω–∞—á–µ–Ω–∏–µ –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é

@router.message(F.text.startswith("http"))
async def handle_download_link(message: Message, state: FSMContext, bot: Bot):
    link = message.text.strip()
    await message.react([ReactionTypeEmoji(emoji="üëÄ")])  # –û—Ç–ø—Ä–∞–≤–ª—è–µ–º —Ä–µ–∞–∫—Ü–∏—é "–≤ –æ–∂–∏–¥–∞–Ω–∏–∏"
    # –ï—Å–ª–∏ —Å—Å—ã–ª–∫–∞ –æ—Ç–Ω–æ—Å–∏—Ç—Å—è –∫ YouTube, –∏—Å–ø–æ–ª—å–∑—É–µ–º –ª–æ–≥–∏–∫—É –≤—ã–±–æ—Ä–∞ –∫–∞—á–µ—Å—Ç–≤–∞
    if "youtube" in link.lower():
        await state.update_data(download_link=link)
        available_qualities = await get_available_qualities(link)
        keyboard = InlineKeyboardBuilder()
        for qual in available_qualities:
            keyboard.button(
                text=f"{qual}",
                callback_data=f"download:{qual}"
            )
        keyboard.adjust(2)
        keyboard.row(
            InlineKeyboardButton(text="Mp3", callback_data="download:mp3")
        )
        await message.answer(
            "–í—ã–±–µ—Ä–∏—Ç–µ –∫–∞—á–µ—Å—Ç–≤–æ –≤–∏–¥–µ–æ –∏–ª–∏ —Å–∫–∞—á–∏–≤–∞–Ω–∏–µ –≤ —Ñ–æ—Ä–º–∞—Ç–µ Mp3:",
            reply_markup=keyboard.as_markup()
        )
        return

    # –î–ª—è –≤—Å–µ—Ö –æ—Å—Ç–∞–ª—å–Ω—ã—Ö —Å—Å—ã–ª–æ–∫ (–Ω–∞–ø—Ä–∏–º–µ—Ä, Instagram, Twitch –∏ –ø—Ä–æ—á–∏—Ö) - —É–ø—Ä–æ—â–µ–Ω–Ω–∞—è –ª–æ–≥–∏–∫–∞
    async def call_download_other():
        try:
            async with grpc.aio.insecure_channel(GRPC_DOWNLOADER_ADDRESS) as channel:
                stub = downloader_pb2_grpc.DownloaderStub(channel)
                request = downloader_pb2.DownloadRequest(
                    url=link,
                    quality="",  # –∫–∞—á–µ—Å—Ç–≤–æ –Ω–µ –≤—ã–±–∏—Ä–∞–µ—Ç—Å—è
                    chat_id=str(message.chat.id),
                    message_id=str(message.message_id),
                    user_id=str(message.from_user.id),
                )
                response = await stub.DownloadVideo(request)
                if response.file_path:
                    if response.file_size > 50 * 1024**2:
                        bot_me = await bot.get_me()
                        full_username = "@" + bot_me.username if bot_me.username else bot_me.first_name
                        await send_file_to_bot(response.file_path, full_username, caption=f"{message.from_user.id}")
                    else:
                        file = await check_path(response.file_path)
                        if file.endswith(".mp4"):
                            await bot.send_video(message.from_user.id, FSInputFile(path=file), caption="–í–∞—à–µ –≤–∏–¥–µ–æ –≥–æ—Ç–æ–≤–æ!")
                        else:
                            await bot.send_audio(message.from_user.id, FSInputFile(path=file), caption="–í–∞—à–µ –∞—É–¥–∏–æ –≥–æ—Ç–æ–≤–æ!")
                        await delete_file(file)
                else:
                    await message.answer("–ó–∞–≥—Ä—É–∑–∫–∞ –∑–∞–≤–µ—Ä—à–∏–ª–∞—Å—å, –Ω–æ —Ñ–∞–π–ª –Ω–µ –Ω–∞–π–¥–µ–Ω.")
        except Exception as e:

            print(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –æ–±—Ä–∞—â–µ–Ω–∏–∏ –∫ —Å–µ—Ä–≤–∏—Å—É –∑–∞–≥—Ä—É–∑–∫–∏: {e}")

    await message.answer("–ó–∞–≥—Ä—É–∑–∫–∞ –Ω–∞—á–∞–ª–∞—Å—å...")
    asyncio.create_task(call_download_other())

@router.callback_query(F.data.startswith("download:"))
async def process_download_choice(callback_query: CallbackQuery, state: FSMContext, bot: Bot):
    data = await state.get_data()
    link = data.get("download_link")
    if not link:
        await callback_query.message.answer("–°—Å—ã–ª–∫–∞ –Ω–µ –Ω–∞–π–¥–µ–Ω–∞, –ø–æ–≤—Ç–æ—Ä–∏—Ç–µ –ø–æ–ø—ã—Ç–∫—É.")
        return
    
    # callback_data –≤ —Ñ–æ—Ä–º–∞—Ç–µ "download:<choice>"
    _, choice = callback_query.data.split(":", maxsplit=1)
    quality = choice if choice != "mp3" else ""
    message_id = str(callback_query.message.message_id)
    user_id = str(callback_query.from_user.id)

    async def call_download():
        try:
            async with grpc.aio.insecure_channel(GRPC_DOWNLOADER_ADDRESS) as channel:
                stub = downloader_pb2_grpc.DownloaderStub(channel)
                if choice == "mp3":
                    request = downloader_pb2.DownloadRequest(
                        url=link,
                        quality="",
                        chat_id=user_id,
                        message_id=message_id,
                        user_id=user_id,
                    )
                    response = await stub.DownloadAudio(request)
                else:
                    request = downloader_pb2.DownloadRequest(
                        url=link,
                        quality=quality,
                        chat_id=user_id,
                        message_id=message_id,
                        user_id=user_id,
                    )
                    response = await stub.DownloadVideo(request)
            if response.file_path:
                if response.file_size > 50 * 1024**2:
                    bot_me = await bot.get_me()
                    full_username = "@" + bot_me.username if bot_me.username else bot_me.first_name
                    await send_file_to_bot(response.file_path, full_username, caption=f"{callback_query.from_user.id}")
                else:
                    file = await check_path(response.file_path)
                    if file.endswith(".mp4"):
                        await bot.send_video(callback_query.from_user.id, FSInputFile(path=file), caption="–í–∞—à–µ –≤–∏–¥–µ–æ –≥–æ—Ç–æ–≤–æ!")
                    else:
                        await bot.send_audio(callback_query.from_user.id, FSInputFile(path=file), caption="–í–∞—à–µ –∞—É–¥–∏–æ –≥–æ—Ç–æ–≤–æ!")
            else:
                await callback_query.message.answer("–ó–∞–≥—Ä—É–∑–∫–∞ –∑–∞–≤–µ—Ä—à–∏–ª–∞—Å—å, –Ω–æ —Ñ–∞–π–ª –Ω–µ –Ω–∞–π–¥–µ–Ω.")
        except Exception as e:
            print(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –æ–±—Ä–∞—â–µ–Ω–∏–∏ –∫ —Å–µ—Ä–≤–∏—Å—É –∑–∞–≥—Ä—É–∑–∫–∏: {e}")

    await callback_query.message.edit_text("–ó–∞–≥—Ä—É–∑–∫–∞ –Ω–∞—á–∞–ª–∞—Å—å...")
    asyncio.create_task(call_download())
    await state.clear()

